### PR TITLE
[RFC]  Property and Filtering Enhancements

### PR DESCRIPTION
**Pull request by jotompki**
*Issue #, if available:* #107, #105, #99,#98, #94, #86, #5, #3

*Description of changes:* Attached is a draft RFC for improvements related to extending key/value access in template to more than just the properties block of a resource type and allowing for more complex filtering of values in collections in templates. Will refine this a bit more next week but want to circulate for now in a draft state. This builds off #111's proposed type block enhancements in some examples. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: rfcs/rfc-113.md [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+```
+AWS::S3::Bucket BucketEncryption.ServerSideEncryptionConfiguration.*.ServerSideEncryptionByDefault
+```
+then make comparisons. If there are more properties that need to be checked in the ServerSideEncryptionByDefault object, this can get verbose quickly. We are proposing enhanced variable assignment from incoming templates. Previously, variables could only be set in the ruleset files themselves and did not have dynamic values. We want to allow customers to do some of these deeply nested checks in an easy fashion by setting a variable to refer to the access, then reusing the variable for a less verbose language.
+
+Assigning parts of the incoming data to variables will allow rule authors to compose simple to read rules:
+```
+%encryption = BucketEncryption.ServerSideEncryptionConfiguration.*.ServerSideEncryptionByDefault
+%encryption.property == "VALUE"
+$encryption.otherProperty IN ["ALLOWED", "VALUES"]
```
Comment (PatMyron) [inline rfcs/rfc-113.md:154]: think I prefer the block-syntax discussed in https://github.com/aws-cloudformation/cloudformation-guard/pull/111 so context can't be split across different parts of rulesets

File: rfcs/rfc-113.md [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+```
+
+With the proposed property access and clause evaluation, one can access any structured data in the input, not only resource type properties. Example ruleset:
+
+```
+Parameters.BucketName EXISTS # ensure a certain parameter exists
+Parameters.BucketName.Type == "String" # conforms to a certain type
+Parameters.BucketName.AllowedValues.* == /\w*bucket\w*/ # make sure values are expected
+
+Resources.*[Type == "AWS::S3::Bucket"].Properties.BucketName == { "Ref": "BucketName"} # ensure bucket name is in allowed values of parameters
```
Comment (PatMyron) [inline rfcs/rfc-113.md:204]: similar to https://github.com/aws-cloudformation/cloudformation-guard/pull/111#discussion_r560576236, I'd really rather not have to specify `Resources` and `Properties` in almost every rule, so I'd want something like the block-syntax if they're both going to be required going forward

File: rfcs/rfc-113.md [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+To disallow all resource KMS Key resource types in a template.
+
+In addition, customers have voiced concerns on the inability to successfully filter resource types matching conditions using the existing "WHEN CHECK" rule type ([#107](https://github.com/aws-cloudformation/cloudformation-guard/issues/107), [#100](https://github.com/aws-cloudformation/cloudformation-guard/issues/100)).  Taken from the above issue, here is the following rule:
+```
+AWS::IAM::Policy WHEN PolicyDocument.Statement.*.Action.* == dynamodb:GetItem CHECK PolicyDocument.Statement.*.Resource != *
+```
+
+and a template: 
+
+```
```
Comment (PatMyron) [inline rfcs/rfc-113.md:91]:
```suggestion
```json
```

File: rfcs/rfc-113.md [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+Note that some of the examples below build off the block evaluations proposed in [#111](https://github.com/aws-cloudformation/cloudformation-guard/pull/111).
+## Proposed enhancements
+
+### Non-resource type scoped Property Access/Support
+
+The non-type scoped property access allows access of any portion of incoming structured data (not just CloudFormation Templates). In the previous iteration of the language, rule authors could only check values of a resource whose type was specified in the rule. Now, one could access any values in the incoming template.
+
+For example, given the CloudFormation template:
+```
```
Comment (PatMyron) [inline rfcs/rfc-113.md:172]:
```suggestion
```json
```

File: rfcs/rfc-113.md (hunk 2 above)
Comment (jotompki) [reply rfcs/rfc-113.md:204]: yeah this was done more to show the functionality of the changes without mixing the two proposals. ideally if these are both merged you could choose how to write these checks: using blocks or some sort of filtering on the top level keys like this

File: rfcs/rfc-113.md (hunk 1 above)
Comment (jotompki) [reply rfcs/rfc-113.md:154]: a lot of this was trying to keep the block evaluations out of this PR as to not be adding those changes to the proposal as well

**Comment by priyap286** [comment]
All issues linked in this RFC PR have been addressed. I am merging this RFC to the Guard1.0 branch.

**Pull request merged**